cloud/scope/azureippool.go

```diff
@@ -0,0 +1,348 @@
+    } else {
+		networkInterface.AcceleratedNetworking = true
+    }
+    return networkInterface
+}
+
+func GetPrimaryNetworkInterface(nics []infrav1.NetworkInterface) (infrav1.NetworkInterface, error) {
+    nic := infrav1.NetworkInterface{}
+	for _,nic := range(nics) {
+        if nic.IsPrimary == true {
+            return nic, nil
+        }
+    }
+    return nic, errors.New("Unable to find a nic with isPrimary set to true")
+}
+
+func AddToAllocatedIPPool(azureippool *infrav1.AzureIPPool, name string, allocatedIP string, vnetName string, subnetName string) *infrav1.AzureIPPool {
+	
+	//iterate through all the allocated IP pools to find if allocated IP pool exists
+	for index, allocatedIPPool := range(azureippool.Status.AllocatedIPPools) {
+		if allocatedIPPool.Name == name {
+			allocatedIPPool.AllocatedIPs = append(allocatedIPPool.AllocatedIPs, allocatedIP)
+			azureippool.Status.AllocatedIPPools[index] = allocatedIPPool
+			return azureippool
+		}
+	}
+
+	allocatedIPPool := InitializeAllocatedIPPool(name, allocatedIP, vnetName, subnetName)
+	azureippool.Status.AllocatedIPPools = append(azureippool.Status.AllocatedIPPools, allocatedIPPool)
+	return azureippool	
+}
+
+func RemoveIPFromAllocatedIPPool(allocatedIPPool infrav1.AllocatedIPPool, freeIP string) (infrav1.AllocatedIPPool, bool) {
+	log := klogr.New()
+	for i, ip := range(allocatedIPPool.AllocatedIPs) {
+		if ip == freeIP {
+			log.Info(fmt.Sprintf("Found IP in allocated ip pool: %s", freeIP))
+			allocatedIPPool.AllocatedIPs = append(allocatedIPPool.AllocatedIPs[0:i],allocatedIPPool.AllocatedIPs[i+1:]...) 
+			return allocatedIPPool, true
+		}
+	}
+	return allocatedIPPool, false
+}
+
+func InitializeAllocatedIPPool(name string, allocatedIP string, vnetName string, subnetName string) infrav1.AllocatedIPPool {
+	allocatedIPPool := infrav1.AllocatedIPPool {
+		Name : name,
+		AllocatedIPs : []string{allocatedIP},
+		VnetName : vnetName,
+		SubnetName : subnetName,
+	}
+	return allocatedIPPool
+}
+
+func GetNextFreeIP(ip string, IPs []string, allocatedIPsMap map[string]bool) (string, error) {
+	log := klogr.New()
+	if ip != "" {
+		isAllocated, exists := allocatedIPsMap[ip]
+		if !exists {
+			return "", errors.New(fmt.Sprintf("IP Address %s not found in the allocatedIPsMap", ip))
+		}
+		if isAllocated {
+			return "", errors.New("APIServerIP %s is marked as allocated.")
+		}
+		log.Info(fmt.Sprintf("Returning ip for controlplane as %s",ip))
+		return ip, nil
+	}
+
+	for _, inputIP := range IPs {
+		isAllocated, exists := allocatedIPsMap[inputIP]
+		if !exists {
+			return "", errors.New(fmt.Sprintf("IP Address %s not found in the allocatedIPsMap", ip))
+		}
+		if !isAllocated {
+			log.Info(fmt.Sprintf("Returning ip for node as %s",inputIP))
+			return inputIP, nil
+		}
+	}
+	return "", errors.New("Failed to get IP for the machine. All IPs are allocated")
+}
+
+func BuildAllocatedIPsMap(azureippool *infrav1.AzureIPPool, ippool infrav1.IPPool) (map[string]bool, error) {
```

Comment:
I think this function signature would be cleaner if you passed in:
BuildAllocatedIPsMap(ippool infrav1.IPPool, allocatedippool infrav1.AllocatedIPPool)

It looks mismatched because you are passing in an IPPool, but, then you have to search azureippool to find the AllocatedIPPool.
Seems like it would be better if you passed in the pair. IPPool and AllocatedIPool

Reply:
Fixing this